### PR TITLE
Add proof support for basic set.filter inferences, add to CPC signature

### PR DESCRIPTION
**Pull request by ajreynol**
Adds proof support for the components responsible for the unsoundness discovered in https://github.com/cvc5/cvc5/pull/11244.

Also adds missing proof support for sets proof rules in CPC signature.

Also fixes cases where proof reconstruction failed for some core sets inferences, fixes https://github.com/cvc5/cvc5-wishues/issues/155.

### CODE REVIEW

File: include/cvc5/cvc5_proof_rule.h [hunk 1]
```diff
@@ -1373,6 +1373,29 @@ enum ENUM(ProofRule)
...
+   * \endverbatim
+   */
+  EVALUE(SETS_FILTER_UP),
+  /**
+   * \verbatim embed:rst:leading-asterisk
+   * **Sets -- Sets filter down**
+   *
+   * .. math::
+   *
+   *   \inferrule{\mathit{set.member}(x,\mathit{set.filter}(P, a))\mid -}
```
Comment (mudathirmahgoub) [inline include/cvc5/cvc5_proof_rule.h:1394]: Is this `\mid -` required? Can't you have a rule with a single premise?

Comment (ajreynol) [reply include/cvc5/cvc5_proof_rule.h:1394]: Mid separates premises from "arguments", which this rule has none.

File: proofs/eo/cpc/rules/Sets.eo [hunk 2]
```diff
@@ -30,3 +30,51 @@
...
+(declare-rule sets_singleton_inj ((T Type) (a T) (b T))
+  :premises ((= (set.singleton a) (set.singleton b)))
+  :conclusion (= a b)
+)
+
+; rule: sets_ext
+; implements: ProofRule::SETS_EXT
+; premises:
+; - deq: A disequality between sets.
+; conclusion: The sets are disequal for a witness lement, as given by the skolem (@set_deq_diff a b).
```
Comment (mudathirmahgoub) [inline proofs/eo/cpc/rules/Sets.eo:48]:
```suggestion
; conclusion: The sets are disequal for a witness element, as given by the skolem (@set_deq_diff a b).
```

File: proofs/eo/cpc/rules/Sets.eo [hunk 3]
```diff
@@ -30,3 +30,51 @@
...
+  :premises ((not (= a b)))
+  :conclusion (not (= (set.member (@sets_deq_diff a b) a) (set.member (@sets_deq_diff a b) b)))
+)
+
+
+; rule: sets_filter_down
+; implements: ProofRewriteRule::SETS_FILTER_DOWN
+; premises:
+; - mem: A membership of element x into a set filter term.
+; conclusion: >
```
Comment (mudathirmahgoub) [inline proofs/eo/cpc/rules/Sets.eo:61]:
```suggestion
; conclusion: 
```

File: proofs/eo/cpc/rules/Sets.eo [hunk 4]
```diff
@@ -30,3 +30,51 @@
...
+;   The membership holds for the set for which we are filtering, and the
+;   predicate is true for x.
+(declare-rule sets_filter_down ((T Type) (P (-> T Bool)) (x T) (S (Set T)))
+  :premises ((set.member x (set.filter P S)))
+  :conclusion (and (set.member x S) (P x))
+)
+
+; rule: sets_filter_up
+; implements: ProofRewriteRule::SETS_FILTER_DOWN
```
Comment (mudathirmahgoub) [inline proofs/eo/cpc/rules/Sets.eo:70]:
```suggestion
; implements: ProofRewriteRule::SETS_FILTER_UP
```

File: proofs/eo/cpc/rules/Sets.eo (hunk 3 above)
Comment (ajreynol) [reply proofs/eo/cpc/rules/Sets.eo:61]: This is used to do multi-line comments based on the doc style.

File: src/theory/sets/theory_sets_private.cpp [hunk 5]
```diff
@@ -730,10 +730,10 @@ void TheorySetsPrivate::checkFilterUp()
...
       Node skolem = d_treg.getProxy(term);
       Node memberFilter = nm->mkNode(Kind::SET_MEMBER, x, skolem);
-      Node not_p_x = p_x.notNode();
-      Node not_memberFilter = memberFilter.notNode();
-      Node orNode =
-          p_x.andNode(memberFilter).orNode(not_p_x.andNode(not_memberFilter));
+      // (set.member x A)
+      // ---------------------------------------
+      // (set.member x (set.filter P A)) = (P x)
+      Node orNode = memberFilter.eqNode(p_x);
```
Comment (mudathirmahgoub) [inline src/theory/sets/theory_sets_private.cpp:736]:
```suggestion
      Node conclusion= memberFilter.eqNode(p_x);
```
Can you update the documentation in header file as well?

**Review by mudathirmahgoub — APPROVED** [review]
LGTM

**Pull request merged**